src/utils/extract.js

```diff
@@ -0,0 +1,96 @@
+const htmlparser = require("htmlparser2")
+
+function parseIndentDescriptor(indentDescriptor) {
+  const match = /^(\+)?(tab|\d+)$/.exec(indentDescriptor)
+  if (!match) {
+    return { relative: false, spaces: "auto" }
+  }
+  return {
+    relative: match[1] === "+",
+    spaces: match[2] === "tab" ? "\t" : Array(Number(match[2]) + 1).join(" "),
+  }
+}
+
+function iterateStyles(code, onStyle) {
+  let index = 0
+  let currentStyle = null
+
+  const parser = new htmlparser.Parser({
+    onopentag(name) {
+      // Test if current tag is a valid <script> tag.
+      if (name !== "style") {
+        return
+      }
+      currentStyle = ""
+    },
+
+    onclosetag(name) {
+      if (name !== "style" || currentStyle === null) {
+        return
+      }
+      onStyle(code.slice(index, parser.startIndex - currentStyle.length), currentStyle)
+      index = parser.startIndex
+      currentStyle = null
+    },
+
+    ontext(data) {
+      if (currentStyle === null) {
+        return
+      }
+      currentStyle += data
+    },
+  })
+
+  parser.parseComplete(code)
+}
+
+export default function extract(code, rawIndentDescriptor) {
+  const indentDescriptor = parseIndentDescriptor(rawIndentDescriptor)
+  const map = []
+  let resultCode = ""
+  let lineNumber = 1
+
+  iterateStyles(code, function (previousCode, stylecode) {
+
+    // Mark that we're inside a <style> a tag and push all new lines
+    // in between the last </style> tag and this <style> tag to preserve
+    // location information.
+    const newLines = previousCode.match(/\r\n|\n|\r/g)
+    if (newLines) {
+      resultCode += newLines.map(function (newLine) {
+        return "/* fake-line */" + newLine
+      }).join("")
+      lineNumber += newLines.length
+      map[lineNumber] = previousCode.match(/[^\n\r]*$/)[0].length
+    }
+
+    const currentStyleIndent = previousCode.match(/([^\n\r]*)<[^<]*$/)[1]
+
+    let indent
+    if (indentDescriptor.spaces === "auto") {
+      const indentMatch = /[\n\r]+([ \t]*)/.exec(stylecode)
+      indent = indentMatch ? indentMatch[1] : ""
+    }
+    else {
+      indent = indentDescriptor.spaces
+      if (indentDescriptor.relative) {
+        indent = currentStyleIndent + indent
+      }
+    }
+
+    resultCode += stylecode
```

Comment:
Please some comments explaining your intentions in this file. There a lot of regular expressions that seem like they might be redundant, but I'm having a hard time figuring out exactly what is supposed to be happening, therefore uncertain what suggestions to make.